frame/staking/src/lib.rs

```diff
@@ -952,11 +952,13 @@ decl_storage! {
 
 		/// The active era information, it holds index and start.
 		///
-		/// The active era is the era currently rewarded.
-		/// Validator set of this era must be equal to `SessionInterface::validators`.
+		/// The active era is the era being currently rewarded. Validator set of this era must be
+		/// equal to [`SessionInterface::validators`].
 		pub ActiveEra get(fn active_era): Option<ActiveEraInfo>;
 
 		/// The session index at which the era start for the last `HISTORY_DEPTH` eras.
+		///
+		/// Note: This tracks the starting session of the [`CurrentEra`].
```

Comment:
I find this comment confusing. This storage tracks the starting session of all era from current_era to current_era - history_depth.

The "starting session" is the index of the session on which the era will start (or has started) validating (i.e. being the active_era)

Reply:
```suggestion
		/// Note: This tracks the starting session (i.e. session index when era start being active) for the eras in `[CurrentEra, CurrentEra - HISTORY_DEPTH]`.
```

Reply:
The point I want to clarify is that we have a scheme like this: 

```
                 +---------+-----------+----------+----------+
    active eras  |   5     |     6     |     7    |     8    |
       +-------+-+-------+-+---------+-+--------+-+--------+-+----------+
  current eras |    5    |     6     |    7     |    8     |
               +---------+-----------+----------+----------+

```

and they are always off by one with the current session system. So the starting session of the active era 6 and starting session of current era 6 are not the same, and `ErasStartSessionIndex` is tracking this for `CurrentEra`.

Reply:
This is why I think generally speaking, the session buffering is not really expressed well through names. `start_session` and `new_session` can both mean the same thing, while one starts an already planned session, and the other one plans a session. 

I recommend a renaming like this:

- new_session -> schedule_session
- start_session -> activate_session
- current_era -> scheduled_era
- active_era -> active_era
- `ErasStartSessionIndex` -> `ScheduledErasStartSessionIndex`


And anything else that I miss about sessions/eras should also clearly say to which category it belongs: `active` stuff or `scheduled` stuff. 

This will need migration, so not now, but if you agree we can make a new issue out of this.

Reply:
> So the starting session of the active era 6 and starting session of current era 6 are not the same

I think this wording doesn't make much sense to me.
`ErasStartSessionIndex` record the session index on which an era have started being active or must start being active.

staking pallet is not aware of how much session is delayed before an era start being active. And staking pallet doesn't track on which session an era become the current era (i.e. staking pallet doesn't track on which session the validator set for an era is chosen). the era 5 may become the current era on session 30 but what ErasStartSessionIndex track is when (i.e. on which session index) the era 5 will start being active (i.e. validating).
if era 5 will start being active on session index 38, and era 5 is current era on session index 30, then `ErasStartSessionIndex(5)==38`

Reply:
Yes having the word schedule instead of current is good to me.

> I recommend a renaming like this:
> 
>    new_session -> schedule_session
    start_session -> activate_session
    current_era -> scheduled_era
    active_era -> active_era
    ErasStartSessionIndex -> ScheduledErasStartSessionIndex

I would rather do `current_era` -> `latest_scheduled_era` or something like this. Because in theory there can be multiple scheduled era ahead of active_era.

`ErasStartSessionIndex -> ErasActivationSessionIndex` maybe ? I don't like ScheduledErasStartSessionIndex because it actually contains information for all era in `latest_scheduled_era to latest_scheduled_era - history_depth`. But yes old eras information is not used. so maybe we can just clean when the era got activated and name `ScheduledErasStartSessionIndex` is fine